thunder/core/utils.py

```diff
@@ -562,8 +562,12 @@ def same_shape(a: Sequence[int], b: Sequence[int], /) -> bool:
 
 
 # TODO: improve error message
-def check_same_shape(*args):
-    shapes = tuple(x.shape for x in args if isinstance(x, TensorProxy))
+# NOTE: the tensors in args have different shapes is only allowed if the tensor is a CPU scalar tensor and allow_cpu_scalar_tensors=True
+def check_same_shape(*args, allow_cpu_scalar_tensors=False):
```

Comment:
same question here, when would we want to assert same device for cpu scalar tensors?

Reply:
Well, I'm not sure, this function is already used in many places, and before this change they all assert the same device for cpu scalar tensors, I saw your comments that this should be the default behavior, but I don't know if I set it to default if any current test cases in Thunder will fail(currently this change only affects binary ops). I'll try it.